com-api/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+//!   description.
+//! - We want to design interfaces such that they're hard to misuse.
+//! - Data communicated over IPC need to be position independent.
+//! - Data may not reference other data outside its provinence
+//! - We provide simple defaults for customization points
+//! - Sending / receiving / calling is always done explicitly.
+//! - COM API does not enforce the necessity for internal threads or executors.
+//!
+//! # Lower layer implementation principles
+//!
+//! - We add safety nets to detect ABI incompatibilities
+//!
+//! # Supported IPC ABI
+//!
+//! - Primitives
+//! - Static lists / strings
+//! - Dynamic lists / strings
+//! - Key-value
+//! - Structures
+//! - Tuples
+
+use std::fmt::Debug;
+use std::ops::{Deref, DerefMut};
+use std::path::Path;
+
+#[derive(Debug)]
+pub enum Error {
+    Fail,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub trait Builder {
+    type Output;
+    /// Open point: Should this be &mut self so that this can be turned into a trait object?
+    fn build(self) -> Self::Output;
+}
+
+pub trait Runtime {}
+
+pub trait RuntimeBuilder: Builder
+where
+    <Self as Builder>::Output: Runtime,
+{
+    fn load_config(&mut self, config: &Path) -> &mut Self;
+}
+
+/// Determines whether a type may be trivially moved without invalidating inner integrity
+pub trait Reloc {}
+
+impl Reloc for () {}
+impl Reloc for u32 {}
+
+/// A `Sample` provides a reference to a memory buffer of an event with immutable value.
+///
+/// By implementing the `Deref` trait implementations of the trait support the `.` operator for dereferencing.
+/// The buffers with its data lives as long as there are references to it existing in the framework.
+pub trait Sample<A, T>: Deref<Target = T> + Send
+where
+    A: Runtime + ?Sized,
+    T: Send + Reloc,
+{
+}
+
+/// A `SampleMut` provides a reference to a memory buffer of an event with mutable value.
+///
+/// By implementing the `DerefMut` trait implementations of the trait support the `.` operator for dereferencing.
+/// The buffers with its data lives as long as there are references to it existing in the framework.
+pub trait SampleMut<A, T>: DerefMut<Target = T>
+where
+    A: Runtime + ?Sized,
+    T: Send + Reloc,
+{
+    /// The associated read-only sample type.
+    type Sample: Sample<A, T>;
+
+    /// Consume the sample into an immutable sample.
+    fn into_sample(self) -> Self::Sample;
+
+    /// Send the sample and consume it.
+    fn send(self) -> Result<()>;
+}
+
+/// A `SampleMaybeUninit` provides a reference to a memory buffer of an event with a `MaybeUninit` value.
+///
+/// Utilizing `DerefMut` on the buffer reveals a reference to the internal `MaybeUninit<T>`.
+/// The buffer can be assumed initialized with mutable access by calling `assume_init` which returns a `SampleMut`.
+/// The buffers with its data lives as long as there are references to it existing in the framework.
+pub trait SampleMaybeUninit<A, T>
```

Comment:
Open question: Should we also add an `impl<A, T> DerefMut for SampleMaybeUninit<A, T>` where `Target=MaybeUninit<T>`?

This would be neat since we could use a lot of methods provided by `MaybeUninit`. However, it would impose the obligation to provide a `MaybeUninit<T>` in the implementation of the actual type (which is imo a must anyway, for correctness reasons).